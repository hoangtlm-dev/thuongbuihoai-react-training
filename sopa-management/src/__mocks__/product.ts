const MOCK_PRODUCTS = [
  {
    name: 'Black shoes',
    price: 87,
    size: ['Small', 'Medium', 'Large'],
    description:
      'Our goal in designing the super-soft Atoms Crew Socks was to provide a seamless match with your Atoms shoes, enhancing your comfort for everyday wear - making them the perfect socks for your everyday shoes.',
    color: 'Black',
    image: {
      url: '/images/products/black-shoes.png',
      alt: 'This is black shoes'
    },
    id: '1'
  },
  {
    name: 'Blue shoes',
    price: 45,
    size: ['Small', 'Medium', 'Large'],
    description:
      'Our goal in designing the super-soft Atoms Crew Socks was to provide a seamless match with your Atoms shoes, enhancing your comfort for everyday wear - making them the perfect socks for your everyday shoes.',
    color: 'Blue',
    image: {
      url: '/images/products/blue-shoes.png',
      alt: 'This is blue shoes'
    },
    id: '2'
  }
];

const MOCK_PRODUCT = {
  name: 'Black shoes',
  price: 87,
  size: ['Small', 'Medium', 'Large'],
  description:
    'Our goal in designing the super-soft Atoms Crew Socks was to provide a seamless match with your Atoms shoes, enhancing your comfort for everyday wear - making them the perfect socks for your everyday shoes.',
  color: 'Black',
  image: {
    url: '/images/products/black-shoes.png',
    alt: 'This is black shoes'
  },
  id: '1'
};

export {
  MOCK_PRODUCTS,
  MOCK_PRODUCT
};
