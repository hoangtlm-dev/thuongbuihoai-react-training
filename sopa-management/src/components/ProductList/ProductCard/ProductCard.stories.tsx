import type { Meta, StoryObj } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';

// Constants
import { IMAGE } from '../../../constants/image';

// Components
import ProductCard from '.';

const meta: Meta<typeof ProductCard> = {
  component: ProductCard,
  decorators: [
    (Story) => (
      <BrowserRouter>
        <Story />
      </BrowserRouter>
    ),
  ],
};

export default meta;
type Story = StoryObj<typeof ProductCard>;

export const TextNormal: Story = {
  args: {
    name: 'Black shoes',
    color: 'Black',
    price: 145,
    href: '/',
    src: IMAGE.blackShoes,
    alt: IMAGE.altBlackShoes,
  }
};
