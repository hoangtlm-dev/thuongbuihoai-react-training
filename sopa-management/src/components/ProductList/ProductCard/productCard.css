.card {
  width: var(--w-medium);
  height: var(--h-large);
  border-radius: var(--rounded-primary);
  background: var(--bg-primary);
  margin-bottom: 30px;
}

.card-image {
  text-align: center;
  margin-top: 80px;
}

.card-image-item {
  width: var(--w-base);
}

.card-item {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  height: var(--h-medium);
  padding: var(--px-medium) var(--py-normal) var(--py-normal) var(--py-normal);
}

.card-description {
  display: flex;
  margin-top: 8px;
  justify-content: space-between;
}

.card-name {
  color: var(--dark);
  font-size: var(--text-base);
  line-height: var(--leading-primary);
  font-weight: var(--font-medium);
  text-decoration: none;
}