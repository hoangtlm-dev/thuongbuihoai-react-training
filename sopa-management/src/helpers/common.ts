// Flattens an array of arrays into a single array
export const flattenArray = <T>(pages: T[][]): T[] => {
  const result: T[] = [];
  pages.forEach((page) => {
    result.push(...page);
  });

  return result;
};
