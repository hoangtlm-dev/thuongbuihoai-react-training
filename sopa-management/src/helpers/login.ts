import {
  Account,
  LoginType
} from "@interfaces/account";
import { FieldError } from "react-hook-form";

export enum CheckType {
  login = 'login',
  email = 'email',
  password = 'password'
}

// Check login
export const checkAccount = (
  data?: Account[],
  email?: string,
  password?: string,
  checkType?: CheckType,
) => {
  switch (checkType) {
    case CheckType.login:
      return data?.some((user: Account) => user.email === email && user.password === password);
    case CheckType.email:
      return data?.some((user: Account) => user.email === email);
    case CheckType.password:
      return data?.some((user: Account) => user.password === password);
    default:
      throw new Error('Invalid check type');
  }
};

export const checkValidationStyles = (
  serverError: boolean,
  rules: FieldError | undefined,
  isDirty: boolean,
  type: LoginType,
) => {
  const checkIsDirty = isDirty ? type.info : type.default
  const style = serverError || rules ? type.error : checkIsDirty;

  return { style };
};