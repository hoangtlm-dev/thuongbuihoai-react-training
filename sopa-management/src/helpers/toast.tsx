import Toast from "../components/Toast";

export const showToast = (
  title: string,
  message: string,
  type: 'success' | 'error',
  duration: number
) => {
  return (
    <Toast
      title={title}
      message={message}
      type={type}
      duration={duration}
    />
  )
};
