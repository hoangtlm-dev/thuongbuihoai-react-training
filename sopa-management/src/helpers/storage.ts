import { Product } from "../interfaces/product";

export const setStorage = (key: string, data: Product) => {
  localStorage.setItem(key, JSON.stringify(data));
};

export const getStorage = (key: string) => {
  return JSON.parse(localStorage.getItem(key) as any) || [];
};

export const clearStorage = () => {
  localStorage.clear();
};
