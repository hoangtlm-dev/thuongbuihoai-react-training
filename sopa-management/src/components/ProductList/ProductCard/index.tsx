import { memo } from "react"
import Text, { SizeType, ThemeType } from "../../common/Text";
import Price, { PriceType } from "../../common/Price";
import Logo from "../../common/Logo";
import './productCard.css';

interface ProductCardProps {
  name?: string;
  color?: string;
  price?: number;
  href: string;
  src: string;
  alt?: string;
};

const ProductCard: React.FC<ProductCardProps> = ({
  name,
  color,
  price,
  href,
  src,
  alt
}) => {
  return (
    <div className="card">
      <Logo href={href} url={src} alt={alt} className="card-image" />
      <div className="card-item">
        <Text text={name} type={SizeType.regular} />
        <div className="card-info">
          <Text text={color} type={ThemeType.highlightPrimary} />
          <Price value={price} type={PriceType.primary} />
        </div>
      </div>
    </div>
  )
}

export default memo(ProductCard);