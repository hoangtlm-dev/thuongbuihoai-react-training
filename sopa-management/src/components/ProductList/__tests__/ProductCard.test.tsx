// Libs
import '@testing-library/jest-dom';

// Components
import ProductCard from '../ProductCard';

// Helpers
import { renderRouterTest } from '../../../helpers/testUtils';

// Mocks
import { MOCK_PRODUCTS } from '../../../__mocks__/product';

describe('ProductList component', () => {
  test('should render ProductList component', () => {
    const { getByTestId } = renderRouterTest(<ProductCard href='/' data={MOCK_PRODUCTS} />);
    expect(getByTestId('product-card')).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const { container } = renderRouterTest(<ProductCard href='/' data={MOCK_PRODUCTS} />);
    expect(container).toMatchSnapshot();
  });
});
