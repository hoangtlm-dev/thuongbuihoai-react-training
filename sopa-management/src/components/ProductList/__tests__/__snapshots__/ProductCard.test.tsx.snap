// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList component matches snapshot 1`] = `
<div>
  <div
    class="product-card"
    data-testid="product-card"
  >
    <div
      class="card"
      data-testid="card"
    >
      <div
        class="card-item"
      >
        <figure
          class="card-image"
        >
          <a
            data-testid="image-link"
            href="/products/1"
          >
            <img
              alt="This is black shoes"
              class="card-image-item"
              src="/images/products/black-shoes.png"
            />
          </a>
        </figure>
        <div
          class="card-info"
        >
          <a
            class="card-name"
            href="/products/1"
          >
            Black shoes
          </a>
          <div
            class="card-description"
          >
            <p
              class=" text-highlight-primary"
              data-testid="text"
            >
              Black
            </p>
            <p
              class=" price-primary"
              data-testid="price"
            >
              $
              87
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="card"
      data-testid="card"
    >
      <div
        class="card-item"
      >
        <figure
          class="card-image"
        >
          <a
            data-testid="image-link"
            href="/products/2"
          >
            <img
              alt="This is blue shoes"
              class="card-image-item"
              src="/images/products/blue-shoes.png"
            />
          </a>
        </figure>
        <div
          class="card-info"
        >
          <a
            class="card-name"
            href="/products/2"
          >
            Blue shoes
          </a>
          <div
            class="card-description"
          >
            <p
              class=" text-highlight-primary"
              data-testid="text"
            >
              Blue
            </p>
            <p
              class=" price-primary"
              data-testid="price"
            >
              $
              45
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
